mainmatter/introduction/pipeline/index.tex

```diff
@@ -0,0 +1,96 @@
+
+The graphics pipeline can be compared to the \gls{CPU} pipeline in the way that it is a number of stages that each command goes through.
+For graphics, that the command is instead an array of vertex data, such as position, texture coordinates, normals, and color, and unlike the \gls{CPU} pipeline, the relationship between the stages are not only one to one.
+
+\begin{figure}[!h]
+  \centering
+  \begin{tikzpicture}[
+      auto, node distance = 2em,
+      every node/.style={font=\footnotesize, align=center},
+      box/.style = {draw, text width=2.3cm, minimum height=3em, drop shadow},
+      fixed stage/.style={box, fill=orange!10},
+      stage/.style={box, rounded corners=1em, fill=blue!10},
+      optional stage/.style={box, fill=blue!5},
+      memory/.style={box, fill=green!10},
+      arrow/.style={draw, ->, thick, >=stealth, shorten >=1pt}]
+      
+    \node[fixed stage](ia){Input\\Assembler};
+    \node[stage, below=of ia](vs){Vertex Shader};
+    \node[fixed stage, below=of vs](pa){Primitive Constructor};
+    \node[optional stage, dashed, below=of pa](t){Tessellation};
+    \node[optional stage, dashed, below=of t](gs){Geometry Shader};
+    \node[fixed stage, below=of gs](r){Rasterizer};
+    \node[stage, below=of r](ps){Fragment Shader};
+    \node[fixed stage, below=of ps](om){Output Merger};
+    
+    \node[memory, fit=(ia)(om), xshift=5cm](m){Memory};
+    
+    \path[arrow] (ia) -- node{Vertex Array} (vs);
+    \path[arrow] (vs) -- node{Vertex Positions} (pa);
+    \path[arrow] (pa) -- node{Primitives} (t);
+    \path[arrow] (t)  -- node{Primitives} (gs);
+    \path[arrow] (gs) -- node{Primitives} (r);
+    \path[arrow] (r)  -- node{Fragment} (ps);
+    \path[arrow] (ps) -- node{Color} (om);
+    
+    \path[arrow] (m.west|-ia) -- (ia);
+    \path[arrow] (m.west|-vs) -- (vs);
+    \path[arrow] (m.west|-pa) -- (pa);
+    \path[arrow] (m.west|-t) -- (t);
+    \path[arrow] (m.west|-gs) -- (gs);
+    \path[arrow] (m.west|-ps) -- (ps);
+    \path[arrow, <->, shorten <=1pt] (m.west|-om) -- (om);
+  \end{tikzpicture}
+  \caption{A view of the graphics pipeline. The orange boxes are fixed functions, the blue boxes with rounded corners are shaders that the developer can replace with their own program, and the large green box is shared memory. Dashed shader stages are optional.}
+  \label{fig:pipeline01}
+\end{figure}
+
+The following paragraph goes through the different stages in the graphics pipeline, show in Figure \ref{fig:pipeline01}.
+
+\paragraph{The Input Assembler} is the first stage. 
+It reads the user provided data and assembles the data into structures that it sends to the next stage in the pipeline.
+These structures include primitive types in the shader languages, such as floats, vectors, matrices, etc.
+
+\paragraph{The Vertex Shader} is the first interchangeable stage in the pipeline.
+Its main responsibility is to take a set of output from the Input Assembler and translate it into vertices in clip-space.
```

Comment:
What's it secondary responsibility? If it doesn't have one, please use a different word.